mainnet-contracts/src/PufferProtocol.sol

```diff
@@ -284,10 +276,87 @@ contract PufferProtocol is IPufferProtocol, AccessManagedUpgradeable, UUPSUpgrad
 
+        // Update numBatches now that validator becomes active
+        $.nodeOperatorInfo[node].numBatches += $.validators[moduleName][index].numBatches;
+
         // Mark the validator as active
         $.validators[moduleName][index].status = Status.ACTIVE;
     }
 
+    /**
+     * @inheritdoc IPufferProtocol
+     * @dev Restricted to Node Operators
+     */
+    function requestConsolidation(bytes32 moduleName, uint256[] calldata srcIndices, uint256[] calldata targetIndices)
+        external
+        payable
+        virtual
+        restricted
+    {
+        if (srcIndices.length == 0) {
+            revert InputArrayLengthZero();
+        }
+        if (srcIndices.length != targetIndices.length) {
+            revert InputArrayLengthMismatch();
+        }
+
+        ProtocolStorage storage $ = _getPufferProtocolStorage();
+
+        bytes[] memory srcPubkeys = new bytes[](srcIndices.length);
+        bytes[] memory targetPubkeys = new bytes[](targetIndices.length);
+        Validator storage validatorSrc;
+        Validator storage validatorTarget;
+        for (uint256 i = 0; i < srcPubkeys.length; i++) {
+            require(srcIndices[i] != targetIndices[i], InvalidValidator());
+            validatorSrc = $.validators[moduleName][srcIndices[i]];
+            require(validatorSrc.node == msg.sender && validatorSrc.status == Status.ACTIVE, InvalidValidator());
+            srcPubkeys[i] = validatorSrc.pubKey;
+            validatorTarget = $.validators[moduleName][targetIndices[i]];
+            require(validatorTarget.node == msg.sender && validatorTarget.status == Status.ACTIVE, InvalidValidator());
+            targetPubkeys[i] = validatorTarget.pubKey;
+
+            // Update accounting
+            validatorTarget.bond += validatorSrc.bond;
+            validatorTarget.numBatches += validatorSrc.numBatches;
+
+            _deleteValidator(validatorSrc);
+            // Node info needs no update since all stays in the same node operator
+        }
+
+        $.modules[moduleName].requestConsolidation{ value: msg.value }(srcPubkeys, targetPubkeys);
+
+        emit ConsolidationRequested(moduleName, srcPubkeys, targetPubkeys);
+    }
+
+    /**
+     * @inheritdoc IPufferProtocol
+     * @dev Restricted to Node Operators
+     */
+    function requestWithdrawal(bytes32 moduleName, uint256[] calldata indices, uint64[] calldata gweiAmounts)
+        external
+        payable
+        restricted
+    {
+        if (indices.length == 0) {
+            revert InputArrayLengthZero();
+        }
+        if (indices.length != gweiAmounts.length) {
+            revert InputArrayLengthMismatch();
+        }
+        ProtocolStorage storage $ = _getPufferProtocolStorage();
+
+        bytes[] memory pubkeys = new bytes[](indices.length);
+
+        // validate pubkeys belong to that node and are active
+        for (uint256 i = 0; i < indices.length; i++) {
+            Validator memory validator = $.validators[moduleName][indices[i]];
```

Comment:
I think that this copies the whole struct to memory, but we only need, `node` `status` `pubkey`, it might look a little uglier, but it might save us some gas for NoOps.
Also, we might not care/need about `status`. If a NoOp calls this for an inactive validator, whats the worst thing that can happen?
No operation tx? User pays for the gas and nothing happens?